pkg/storegateway/sharding_strategy.go

```diff
@@ -110,6 +118,21 @@ func (s *ShuffleShardingStrategy) FilterBlocks(_ context.Context, userID string,
 		return nil
 	}
 
+	// If this user has been explicitly disabled or _other_ users have been explicitly
+	// enabled and this user isn't, don't sync any blocks for this particular user. This
+	// should not happen in practice since the user would have been filtered before this
+	// method is called but, we check just in case.
+	if !s.allowedTenants.IsAllowed(userID) {
+		level.Warn(s.logger).Log("msg", "user is disabled by configuration but block sync is still being attempted", "user", userID)
+
+		for blockID := range metas {
+			synced.WithLabelValues(shardExcludedMeta).Inc()
+			delete(metas, blockID)
+		}
+
+		return nil
+	}
```

Comment:
This seems unnecessary, and signals that maybe `ShuffleShardingStrategy` is not a good place for checking for allowed users. Have you considered moving check for allowed users to `BucketStores` (e.g. `scanUsers`)? Then we wouldn't need the check here.

Reply:
Updated. I've also changed the logic of `scanUsers` to _not_ return partial results in case of an error. The method was only called in two places and both places made of a point of not using the partial results.